compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp

```diff
@@ -0,0 +1,156 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Util/Analysis/Explorer.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/GlobalOptimization/PassDetail.h"
+#include "iree/compiler/GlobalOptimization/Passes.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Utils/ReshapeOpsUtils.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir::iree_compiler::GlobalOptimization {
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Util functions
+//===----------------------------------------------------------------------===//
+
+static LogicalResult
+foldUnitDimsOnGlobal(IRRewriter &rewriter, IREE::Util::GlobalOpInterface global,
+                     SmallVector<IREE::Util::GlobalLoadOpInterface> loadOps,
+                     SmallVector<IREE::Util::GlobalStoreOpInterface> storeOps,
+                     SymbolTable moduleSymbols) {
+  // Create a new transformed GlobalOp.
+  SmallVector<int64_t> newShape;
+  auto globalType = cast<RankedTensorType>(global.getGlobalType());
+  for (auto size : globalType.getShape()) {
+    if (size != 1) {
+      newShape.push_back(size);
+    }
+  }
+  // Just bail in the special case for tensor of all unit dims.
+  if (newShape.empty()) {
+    return success();
+  }
+  auto newGlobalType =
+      RankedTensorType::get(newShape, globalType.getElementType());
+  std::optional<TypedAttr> newInitialValue = std::nullopt;
```

Comment:
The init value is `std::nullopt`, so you dont have to specify it.

https://en.cppreference.com/w/cpp/utility/optional/optional